### ISSUE TITLE
[Python] [FlightRPC] Suspicious amounts of memory held on client after failed DoExchange

### ISSUE THREAD

**Issue by lupko** [apache/arrow#36540]
### Describe the bug, including details regarding any error messages, version, and platform.

When a PyArrow FlightClient calls do_exchange and the call fails on server and the client holds onto the raised `FlightError`, then it seems that not all memory that was used during do_exchange processing is returned.

---

Let me provide some more context to paint a better picture. In our system, we have a Flight RPC service that can generate data using Flight Descriptor containing a command. The command specifies a flight path of source data, flight path where to sink the data + some payload for the processing. 

The command processing may take a while - it is a long running command and the service represents this internally as a task that it queues up. The clients 'submit' these tasks via GetFlightInfo and keep polling on the result.

When the task runs, it needs to do some expensive computation: this is what we have to offload to a worker process via DoExchange. The worker process has its own Flight RPC server (bound to unix socket).

The task will start DoExchange with the worker process, then open stream of the source data and shovel it to the worker. At this point, the worker does its work -> which may fail. On failure, the main process captures the error and makes it a result of the task - so that when client comes polling for the task result, they learn that it failed with that particular error.

What we found during load testing error scenarios is, that the RSS of the server (e.g. the main process) only ever goes up.

Note that there is no funny stuff on success. The worker produces data, the main process does DoPut for the sink flight path and streams the data where necessary. Memory usage in this case is as expected.

---

After a lot of head scratching I think I have finally pin-pointed this suspicious memory usage to the fact that the main process holds onto the exception that failed the DoExchange. If the code does not keep the exception, the memory usage is ok. As soon as it holds onto it, the memory accumulates. The rate of accumulation seems to correlate with the size of the source data.

---

Here is a gist with the reproducer that I cut to the bare-bones: https://gist.github.com/lupko/ce59226b99790af2474bfe459cb4e36d

The reproducer does `malloc_trim` to drop any unused memory that malloc has not returned to the system. You will be able to observe that the trim with accumulated errors does not impact process RSS at all. Only when the accumulated errors are dropped, the memory usage goes down (significantly).

### Component(s)

FlightRPC, Python

**Comment by lidavidm** [comment]
Hmm, interesting, thanks for investigating. I wonder if we're accidentally capturing something in a closure or similar...

**Comment by lidavidm** [comment]
Oh, but it's the RSS of the _server_ that goes up?

**Comment by lidavidm** [comment]
> errors in `to_reader()` call are propagated as OSError (suspicious?)

This is tangential, but IMO this was a mistake of the original implementation of trying to reuse the Arrow errors instead of specifying a fully independent set of Flight-specific errors. So Arrow converts things like IOError (status code) to OSError (exception), and Flight reuses that (and additionally tries to map the gRPC status codes onto the Arrow status codes), which gets confusing fast.

**Comment by lupko** [comment]
hello @lidavidm, thank you for checking out the issue. to clarify:

- the server - `repro_server.py` - the one that handles the DoExchange is fine. 
- it is the client - `repro_client.py` - the one that initiates the DoExchange - where the RSS keeps accumulating. 

sorry if my description was a bit confusing... in our case the client who initiates the do_exchange is actually also a server :) 

but, long story short, you can see the increased RSS in the process that runs the `repro_client.py` from the reproducer. the `repro_server.py` is fine (sans the usual malloc/malloc trim stuff).

**Comment by lidavidm** [comment]
Ok, thanks for clarifying!

I haven't gotten to investigate yet, sorry. I'll report back once I can.

**Comment by lupko** [comment]
i did some further digging and turns out this is related to Python exception handling. The stack frames included in exception's traceback keep reference to all local variables. And so naturally things do not get cleared up if code holds onto the exception.

There is even a function `traceback.clear_frames(tb)` to address this. Tested it with the reproducer and all is good.

So if anything, this could perhaps be a little doc update? A note into `pyarrow.flight.FlightError` perhaps? WDYT?

Otherwise I'm going to close this.

**Comment by lidavidm** [comment]
Ok, not surprising.

If possible I'd like to keep this open so we can figure out what is being held open, and if we can clear it explicitly instead of just documenting it.

**Comment by lupko** [comment]
of course, although I suspect there is very little that could be done on PyArrow side - it all looks like a Python thing to me:

For example if I fiddle-up the existing reproducer (`repro_client.py`) so that the `_do_exchange` method raises any error from Python-land and I hold onto that error, then I observe the same symptoms: because the exception references the traceback which references the stack frames and one of them references all local variables of `_do_exchange` - even those that are backed by instances of Arrow classes.

FYI here is some extra info about the exception funny stuff: https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/
